### ISSUE TITLE
Incorrectly separating imports

### ISSUE THREAD

**Issue by japgolly** [liancheng/scalafix-organize-imports#59]
My `.scalafix.conf` is:

```
rules = [
  OrganizeImports,
  RemoveUnused
]

RemoveUnused {
  imports  = false
  privates = true
  locals   = true
}

OrganizeImports {
  expandRelative       = true
  groupedImports       = Merge
  groups               = ["*"]
  importSelectorsOrder = Ascii
  removeUnused         = true
}
```

and when I format this:

```scala
import cats.CommutativeApplicative
import cats.effect.IO
import cats.effect.IO.ioEffect
import java.time.{Duration, Instant}
import scala.collection.Factory
import scala.concurrent.blocking
import scalaz.{-\/, BindRec, Catchable, Monad, \/, \/-}
```

it turns it into this:

```scala
import cats.CommutativeApplicative
import cats.effect.IO
import java.time.{Duration, Instant}
import scala.collection.Factory
import scala.concurrent.blocking
import scalaz.{-\/, BindRec, Catchable, Monad, \/, \/-}

import cats.effect.IO.ioEffect
```

It should do nothing according to my config.

**Comment by japgolly** [comment]
Another strange case of it separating imports:

```scala
import japgolly.microlibs.nonempty.NonEmptyVector
import japgolly.univeq.UnivEq
import scala.annotation.tailrec
import scalaz.{Applicative, Equal, Order, Traverse}
import scalaz.std.map.mapEqual
import scalaz.std.option.optionInstance
import scalaz.syntax.traverse._
```

is being turned into

```scala
import japgolly.microlibs.nonempty.NonEmptyVector
import japgolly.univeq.UnivEq
import scala.annotation.tailrec
import scalaz.syntax.traverse._
import scalaz.{Applicative, Equal, Order, Traverse}

import scalaz.std.map.mapEqual
import scalaz.std.option.optionInstance
```

**Comment by japgolly** [comment]
Yet another strange case, this time a bit worse because it's splitting one line into three

```scala
import japgolly.univeq.UnivEq
import scalaz.Order
import scalaz.std.string.stringInstance
import scalaz.std.anyVal.{intInstance, longInstance, shortInstance}
```

is being turned into

```scala
import japgolly.univeq.UnivEq
import scalaz.Order

import scalaz.std.string.stringInstance
import scalaz.std.anyVal.intInstance
import scalaz.std.anyVal.longInstance
import scalaz.std.anyVal.shortInstance
```

**Comment by japgolly** [comment]
FYI I've tested this with both 0.3.0 and 0.3.1-RC1 of this plugin, with sbt-scalafix 0.9.15{,-3}, semanticDB 4.3.10, and Scala 2.13.2.

**Comment by liancheng** [comment]
@japgolly, first of all, you don't want to use `OrganizeImports` together with `RemoveUnused`. Instead, you may set `OrganizeImports.removeUnused` to `true`. That's probably one of the reasons why you ended up with weird results. Please refer to [this paragraph][1] for details.

The 2nd reason that caused part of the unexpected results is probably because of explicitly imported implicits. See the "explicitly imported implicit names" section of the documentation of the [`groups` option][2].

[1]: https://github.com/liancheng/scalafix-organize-imports/tree/6ccfe5ede129a06dbfbeac13ec17b7b4973f18aa#remove-unused-warning
[2]: https://github.com/liancheng/scalafix-organize-imports/tree/6ccfe5ede129a06dbfbeac13ec17b7b4973f18aa#25-groups

**Comment by liancheng** [comment]
Some related design decisions were discussed in issue #40. Please let me know what you think about it. Maybe I should provide an IntelliJ compatible mode even if that may sacrifice correctness for some rare edge cases.

For the explicitly imported implicit names case, using a wildcard import should be good enough to avoid it. I thought in most cases people simply use wildcard imports to import implicits, but maybe I'm wrong.

**Comment by liancheng** [comment]
Oh, and about `RemoveUnused`, you can still use it to remove unused private fields and local variables. Just don't use `RemoveUnused` to remove unused imports when `OrganizeImports` is also invoked.

**Comment by liancheng** [comment]
Another thing I wanted to try is to add a comment above the last order-preserving group containing relative imports and implicits. Something like:

```scala
// Order-preserving import group for relative imports and explicitly imported implicit names.
```

And maybe provide an option to allow users to remove this comment. Might not be a good idea, though...

**Comment by japgolly** [comment]
Hey. Yeah I read that documentation, that's why in my settings I've set `RemoveUnused.imports  = false`. Are you saying that `RemoveUnused.imports  = false` isn't enough, people should remove the `RemoveUnused` entirely?

**Comment by liancheng** [comment]
Ah, sorry, @japgolly, my bad! Misread your configuration. Then this does seem to be a bug, let me take a look. As always, thanks for reporting!

**Comment by japgolly** [comment]
Ahh I see this:

> To avoid having such implicit imports appearing before any wildcard import that may introduce a conflicting implicit, they are always moved to the last import group.

I didn't see that before. I'd love if there was an option to disable this behaviour and ignore this complication. (Personally I don't think anyone in the Scala community should be tolerating it, but that's just my opinion).

> For the explicitly imported implicit names case, using a wildcard import should be good enough to avoid it. I thought in most cases people simply use wildcard imports to import implicits, but maybe I'm wrong.

Yeah I often import only specify implicits. It's common across all of my codebases and I've seen others doing it too -- I'm not alone -- but I have no idea what the ratio of always-wildcard vs specifics is at the community level.

**Comment by japgolly** [comment]
> Ah, sorry, @japgolly, my bad! Misread your configuration. Then this does seem to be a bug, let me take a look. As always, thanks for reporting!

hehe no worries and thanks so much for both looking into and everything you've done with this plugin. Really it's fantastic!

**Comment by liancheng** [comment]
Yeah, also confirmed that all the affected imports are implicit names. And breaking one grouped imports consisting of implicit names only into multiple imports is indeed unexpected and need to be fixed.

Alright, then I guess I should bring back PR #42 (merged but later reverted) to introduce an option to disable this feature.

**Comment by liancheng** [comment]
@sideeffffect, you got a comrade :) Now I'm convinced.

**Comment by japgolly** [comment]
> Alright, then I guess I should bring back PR #42 (merged but later reverted) to introduce an option to disable this behavior.

That would be wonderful! Only suggestion I have is I think maybe a different name for the setting would be better. I don't want this behaviour in order to be compatible with IntelliJ, I just think it's a very (*very*) bad idea to tolerate and rely on the order of imports at the same level. If I ever accidentally have that case in my code I want to fix it rather than be very cafeful about import order (for myself and all future authors of the file). If we call it an Intellij-mode then I'm worried that in future maybe some more behaviour is put behind the same config option and then users don't have the ability to fully control the both behaviours (because it becomes none or all). What would you think about a name like "preserveImplicitOrder" or "carefulImplicitOrdering" or something similar?

**Comment by japgolly** [comment]
Oh and btw, just because it's interesting, let me tell you that there's another reason that I sometimes choose to import a specific implicit value instead of a wildcard: when I'm working on Scala.JS code I have to be careful about what all I'm depending on because everything reachable is included in the output JS when users then download. If you import a wildcard then sometimes you end up using something you didn't expect, or maybe after some modifications: something you thought you stopped using, and it could pull in a bunch of other stuff and suddenly your JS is huge. Making it explicit just gives you a bit more control and confidence in certain situations. Obviously I don't do it everywhere but sometimes it adds value :)

**Comment by liancheng** [comment]
> Only suggestion I have is I think maybe a different name for the setting would be better.

Yeah... The hardest computer science problem: naming. I struggled a lot while naming this option. And part of the reason why I reverted it later was that I didn't like the name. Let me think about it. Your rationale makes a lot of sense.

Maybe I should not do any special casing. Instead, just document this weird edge case and tell the readers: you should never code in this way. If your code is ever broken by `OrganizeImports` due to this, it's really your own fault...

> If you import a wildcard then sometimes you end up using something you didn't expect, or maybe after some modifications: something you thought you stopped using, and it could pull in a bunch of other stuff and suddenly your JS is huge. Making it explicit just gives you a bit more control and confidence in certain situations.

This makes perfect sense. Then I do expect explicitly importing an implicit to be a reasonable practice. In fact, personally I'm also follow this practice for no good reasons (`OrganizeImports` is [not an exception either][1]). But I thought I was just weird...

[1]: https://github.com/liancheng/scalafix-organize-imports/blob/6ccfe5ede129a06dbfbeac13ec17b7b4973f18aa/rules/src/main/scala/fix/OrganizeImports.scala#L29-L30

**Comment by liancheng** [comment]
Marked this issue as a duplicate of #40.

**Comment by liancheng** [comment]
Fixed in #61 and made available in the newly cut 0.3.1-RC2 release.
